### PR TITLE
Add tasks scheduler

### PR DESCRIPTION
**Pull request by Giost**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] PR has a descriptive title
- [x] PR contains closely related changes only (should not have many unrelated changes)
- [x] STEP pod members have been added as reviewers
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the new feature?

The scheduler receives a list of calendar events and tasks and it schedules the tasks in free time slot of the calendar.

### CODE REVIEW

File: src/main/java/com/google/sps/scheduler/Scheduler.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+   * all of them is returned.
+   */
+  public static List<Event> schedule(List<Event> calendarEvents, String timeZone, LocalDate dayDate) {
+    return schedule(calendarEvents, TASKS_SAMPLE, timeZone, dayDate);
+  }
+
+  /**
+   * Returns an event with the given start and end time in the specific time zone.
+   */
+  public static Event createEvent(DateTime startTime, DateTime endTime, String timeZone) {
```
Comment (Giost) [inline src/main/java/com/google/sps/scheduler/Scheduler.java:121]: Probably this method and the following should be moved to the class that manages the calendar API

File: src/main/java/com/google/sps/scheduler/Scheduler.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      }
+
+      while (eventStart - lastEnd >= DEFAULT_DURATION_IN_MILLISECONDS &&
+          indexNextTask < tasks.size()) {
+        DateTime startTime = epochToDateTime(lastEnd, timeZone);
+        lastEnd += DEFAULT_DURATION_IN_MILLISECONDS;
+        DateTime endTime = epochToDateTime(lastEnd, timeZone);
+        tasksEvents.add(
+            createEventWithSummary(startTime, endTime, timeZone, tasks.get(indexNextTask)));
+        indexNextTask++;
```
Comment (dola) [inline src/main/java/com/google/sps/scheduler/Scheduler.java:83]: you can instead us an [Iterator](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html) which you get by calling [`tasks.iterator()`](https://docs.oracle.com/javase/8/docs/api/java/util/List.html#iterator--)

File: src/main/java/com/google/sps/scheduler/Scheduler.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  /**
+   * Schedules some sample tasks in the free time slot of the calendar events, which must be of
+   * the same day specified in the parameter.
+   * For each task that can be scheduled, a new calendar event object is created and a list of
+   * all of them is returned.
+   */
+  public static List<Event> schedule(List<Event> calendarEvents, String timeZone, LocalDate dayDate) {
+    return schedule(calendarEvents, TASKS_SAMPLE, timeZone, dayDate);
+  }
```
Comment (dola) [inline src/main/java/com/google/sps/scheduler/Scheduler.java:116]: This and the referenced fake tasks data should instead be in the test code as it is unused in production.

File: src/main/java/com/google/sps/scheduler/Scheduler.java (hunk 1 above)
Comment (zlorf) [reply src/main/java/com/google/sps/scheduler/Scheduler.java:121]: +1

File: src/main/java/com/google/sps/scheduler/Scheduler.java (hunk 3 above)
Comment (Giost) [reply src/main/java/com/google/sps/scheduler/Scheduler.java:116]: They are used in the prototype to show a tasks list in the UI.

Comment (dola) [reply src/main/java/com/google/sps/scheduler/Scheduler.java:116]: I see. Maybe wen can then add a TODO (potentially referencing a Github issue) to the code to move this out. Alternatively you could add a class already for the Tasks integration that is coming up and let it return the fake data from there so the separation is already given.

File: src/main/java/com/google/sps/scheduler/Scheduler.java [hunk 4]
```diff
@@ -37,6 +38,7 @@
   public static final int END_HOUR = 18;
   public static final int END_MINUTE = 0;
   public static final long DEFAULT_DURATION_IN_MILLISECONDS = TimeUnit.MINUTES.toMillis(30);
+  // TODO: remove the sample tasks after the Tasks API integration (issue #8)
   public static final ImmutableList<String> TASKS_SAMPLE = ImmutableList.of(
```
Comment (zlorf) [inline src/main/java/com/google/sps/scheduler/Scheduler.java:42]: Sorry for being late to the party, but can these sample tasks be at least moved to some other file? I imagine you'd like to show them in the UI, and accessing them from Scheduler does not seem right.

Comment (Giost) [reply src/main/java/com/google/sps/scheduler/Scheduler.java:42]: Yes, it makes sense.

**Review by fuiszgt — APPROVED** [review]

**Review by zlorf — APPROVED** [review]

**Review by dola — APPROVED** [review]

**Pull request merged**